nncf/nncf_network.py

```diff
@@ -74,34 +74,49 @@ def __eq__(self, other):
 
 class InsertionInfo:
     def __init__(self, op_exec_context: OperationExecutionContext,
+                 in_port_id: Optional[int] = None,
                  is_input=False,
                  is_output=False,
                  shape_to_operate_on=None):
         self.op_exec_context = op_exec_context  # type: OperationExecutionContext
+        self.in_port_id = in_port_id  # None for post-hook quantization, otherwise - pre-hook
         self.is_input = is_input
         self.is_output = is_output
         self.shape_to_operate_on = shape_to_operate_on
-        self.linked_op_exec_contexts = []  # type: List[OperationExecutionContext]
+        self.linked_insertion_infos = []  # type: List[InsertionInfo]
 
     def __eq__(self, other: 'InsertionInfo'):
-        return self.op_exec_context == other.op_exec_context and Counter(self.linked_op_exec_contexts) == Counter(
-            other.linked_op_exec_contexts)
+        # TODO: ensure no circular refs via self.linked_insertion_infos?
+        return self.op_exec_context == other.op_exec_context and Counter(self.linked_insertion_infos) == Counter(
+            other.linked_insertion_infos) and self.in_port_id == other.in_port_id
+
+    def __str__(self):
+        postfix = ''
+        if self.in_port_id is not None:
+            postfix = "|INPUT{}".format(self.in_port_id)
+        return str(self.op_exec_context.input_agnostic) + postfix
 
     def __hash__(self):
-        return self.op_exec_context.__hash__()
+        return hash(str(self))
 
 
 class InsertionPoint:
     def __init__(self, ia_op_exec_context: InputAgnosticOperationExecutionContext,
-                 insertion_type: InsertionType):
+                 insertion_type: InsertionType,
+                 input_port_id: int = None):
         self.ia_op_exec_context = ia_op_exec_context
```

Comment:
An insertion point now stores an input port to affect, in order to represent the pre-hook semantics as well. `None` specifies a post-hook insertion.